### DISCUSSION TITLE
How to implement OpenTelemetry tracing

### DISCUSSION THREAD

**Discussion by tadeas-vintrlik** [buildbot/buildbot#7771] [Q&A]
I was wondering if you might have an idea what is the best way to implement observability using OpenTelemetry (or other).

From what I know OpenTelemetry has python library, I checked it out and you can create trace spans either by directly creating a span object or by using a decorator on the function you wish to trace.

Since in the buildbot config you just register the builders I do not think the decorator would be helpful there. One way I came with would be to create a consumer for the builds and steps message queue but I am not sure this is the best approach.

Does anyone have any recommendations?

**Comment by chucklever** [comment]
> I was wondering if you might have an idea what is the best way to implement observability using OpenTelemetry (or other).
> 
> From what I know OpenTelemetry has python library, I checked it out and you can create trace spans either by directly creating a span object or by using a decorator on the function you wish to trace.

Newcomer here. I've been looking at OpenTelemetry for Linux's NFS stack, and I wasn't aware that Python had a library for OTEL.

One question for you: can you elaborate on what, in particular, you want to observe? For example, are you looking for ways to observe interactions between the build master and the workers?

Signed, Curious Bystander

**Reply by tadeas-vintrlik** [reply]
Basically I would like to be able to better observe the builds and specifically their steps and locks. We utilize locks quite heavily and it would be nice to have a clear overview of what is happening. Which builds run simultaneously, which are waiting for resources, etc.

But in general we are just trying to monitor Buildbot from Grafana, which is the central place for all our observability. We are using an in-house Prometheus exporter I've written, but this is quite a stretch for Prometheus so I am looking for other options and OTEL seemed best suited for CI/CD monitoring.
